pg-extend/src/pg_datum.rs

```diff
@@ -370,6 +363,101 @@ where
     }
 }
 
+/// Inner trait used to limit which types can be used for direct casting
+#[doc(hidden)]
+pub trait PgPrimitiveDatum {}
+
+impl PgPrimitiveDatum for i16 {}
+impl PgPrimitiveDatum for i32 {}
+impl PgPrimitiveDatum for i64 {}
+impl PgPrimitiveDatum for f32 {}
+impl PgPrimitiveDatum for f64 {}
+
+impl<'s, T> TryFromPgDatum<'s> for &[T]
+where
+    T: 's + TryFromPgDatum<'s> + PgPrimitiveDatum,
+{
+    fn try_from<'mc>(_: &'mc PgAllocator, datum: PgDatum<'mc>) -> Result<Self, &'static str>
+    where
+        Self: 's,
+        'mc: 's,
+    {
+        if let Some(datum) = datum.0 {
+            unsafe {
+                let datum = datum as *mut pg_sys::varlena;
+                if datum.is_null() {
+                    return Err("datum was NULL");
+                }
+
+                #[allow(clippy::cast_ptr_alignment)]
+                let arr_type = pg_sys::pg_detoast_datum(datum) as *mut pg_sys::ArrayType;
+
+                if (*arr_type).ndim > 1 {
+                    return Err("argument must be empty or one-dimensional array");
+                }
+
+                let mut elmlen: pg_sys::int16 = 0;
+                let mut elmbyval = pgbool!(false);
+                let mut elmalign: ::std::os::raw::c_char = 0;
+
+                pg_sys::get_typlenbyvalalign(
+                    (*arr_type).elemtype,
+                    &mut elmlen,
+                    &mut elmbyval,
+                    &mut elmalign,
+                );
+
+                let mut nulls = std::ptr::null_mut::<pg_sys::bool_>();
+                let mut elements = std::ptr::null_mut::<Datum>();
+                let mut nelems: i32 = 0;
+
+                pg_sys::deconstruct_array(
```

Comment:
does this allocate data potentially?

Reply:
It can, especially if the Datum comes from disk.

Not so much from direct SQL statements that pass in `ARRAY[1,2,3]`

Reply:
Actually, I lied,  deconstruct_array() will always allocate memory.  It allocates an array to store the returned elements, along with an array to store the null element bitmap.

If there's also a call to `pg_detoast_datum()`, that's the thing that *might* allocate, depending on the Datum's storage characteristics.

Reply:
There is a call to `pg_detoast_datum()`. Since this implementation is limited to Datums of type int16/int32/int64/f32/f64, would it allocate for those? Or does it depends on something elese?